### PR TITLE
Make ensureGreen and ensureYellow wait for cluster size consistency

### PR DESCRIPTION
**Pull request by ywelsch**
We currently often use ensureGreen or ensureYellow to check whether the cluster is in a good state again after shutting down a node:
```
stopRandomNode(...);
ensureGreen() / ensureYellow();
```

With the change in #21092, however, it can happen that if the node that is stopped is the master node, another node will become master and publish a cluster state where it is master but where the node that was stopped hasn't been removed yet from the cluster state. It will only publish a second state thereafter where the old master is removed. If the ensureGreen/ensureYellow is timed just right, it will get to execute before the second cluster state update removing the old master and the condition ensureGreen / ensureYellow might not hold at that point anymore.

### CODE REVIEW

File: test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java [hunk 1]
```diff
@@ -869,15 +869,45 @@ public ClusterHealthStatus ensureGreen(String... indices) {
...
+        ClusterHealthRequest healthRequest = Requests.clusterHealthRequest(indices)
+            .timeout(timeout)
+            .waitForEvents(Priority.LANGUID)
+            .waitForNoRelocatingShards(true);
+        if (green) {
+            healthRequest.waitForGreenStatus();
+        } else {
+            healthRequest.waitForYellowStatus();
+        }
+        if (cluster() != null) {
```
Comment (bleskes) [inline test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java]: why is this checked? `client()` needs a cluster to be there, so it should always be non null?

File: test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java [hunk 2]
```diff
@@ -869,15 +869,45 @@ public ClusterHealthStatus ensureGreen(String... indices) {
...
+    private ClusterHealthStatus ensureGreenOrYellow(boolean green, TimeValue timeout, String... indices) {
+        String color = green ? "green" : "yellow";
+        String method = green ? "ensureGreen" : "ensureYellow";
+
+        ClusterHealthRequest healthRequest = Requests.clusterHealthRequest(indices)
+            .timeout(timeout)
+            .waitForEvents(Priority.LANGUID)
+            .waitForNoRelocatingShards(true);
+        if (green) {
+            healthRequest.waitForGreenStatus();
```
Comment (bleskes) [inline test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java]: there is a waitForStatus variant  - should be simpler.

File: test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java [hunk 3]
```diff
@@ -869,15 +869,45 @@ public ClusterHealthStatus ensureGreen(String... indices) {
...
+        } else {
+            healthRequest.waitForYellowStatus();
+        }
+        if (cluster() != null) {
+            healthRequest.waitForNodes(Integer.toString(cluster().size()));
```
Comment (bleskes) [inline test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java]: can you add a comment as why we do this?

**Review by bleskes — CHANGES_REQUESTED** [review]
Left some very minor suggestions. I'm still on the fence if this should be done here or as a functionality of the test cluster (shut down a node -> wait for that to be processed too), where people may opt out of. Let's get this in as is and see _if_ we want to change in the future.

**Comment by ywelsch** [comment]
@bleskes I've pushed a change addressing review comments.

File: test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java [hunk 4]
```diff
@@ -904,7 +906,7 @@ private ClusterHealthStatus ensureGreenOrYellow(boolean green, TimeValue timeout
                 client().admin().cluster().preparePendingClusterTasks().get());
             fail("timed out waiting for " + color + " state");
         }
-        if (green) {
+        if (clusterHealthStatus == ClusterHealthStatus.GREEN) {
```
Comment (bleskes) [inline test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java]: nit: we can change this to greater than or equal relationship and check on both statuses.

**Review by bleskes — APPROVED** [review]
LGTM. Thx @ywelsch

**Comment by ywelsch** [comment]
thanks @bleskes

**Comment by ywelsch** [comment]
backported to 5.2.0

**Pull request merged**